app/scripts/modules/core/pipeline/config/copyStage/copyStage.modal.controller.ts

```diff
@@ -0,0 +1,96 @@
+import {flatten} from 'lodash';
+import {module} from 'angular';
+
+import {API_SERVICE, Api} from 'core/api/api.service';
+import {Application} from 'core/application/application.model';
+import {COPY_STAGE_CARD_COMPONENT} from './copyStageCard.component';
+import {IPipeline} from 'core/domain/IPipeline';
+import {IStage} from 'core/domain/IStage';
+
+import './copyStage.modal.less';
+
+interface IStageWrapper {
+  pipeline: string;
+  stage: IStage;
+}
+
+class CopyStageModalCtrl implements ng.IComponentController {
+  viewState = { loading: true, error: false };
+  applications: Application[];
+  stages: IStageWrapper[];
+  selectedStage: IStageWrapper;
+
+  static get $inject() { return ['$q', 'API', 'application', 'applicationReader', '$uibModalInstance']; }
+
+  constructor (public $q: ng.IQService,
+               public API: Api,
+               public application: Application,
+               public applicationReader: any,
+               public $uibModalInstance: any) { }
+
+  $onInit (): void {
+    this.$q.all({
+      stages: this.getStagesForApplication(this.application.name),
+      applications: this.applicationReader.listApplications(),
+    })
+    .then(({ stages, applications }) => {
+      this.stages = stages;
+      this.applications = applications;
+      this.viewState.loading = false;
+      this.viewState.error = false;
+    })
+    .catch(() => {
+      this.viewState.loading = false;
+      this.viewState.error = true;
+    });
+  }
+
+  getStagesForApplication (applicationName: string): ng.IPromise<IStageWrapper[]> {
+    return this.API.one('applications')
+      .one(applicationName)
+      .all('pipelineConfigs')
+      .getList()
+      .then((pipelines: IPipeline[]) => {
+        let nestedStageWrappers = pipelines
+          .map((pipeline) => {
+            return (pipeline.stages || [])
+              .map((stage: IStage) => {
+                return { pipeline: pipeline.name, stage: stage };
+              });
+          });
+
+        return flatten(nestedStageWrappers);
+      });
+  }
+
+  onApplicationSelect (selected: Application): void {
+    this.viewState.loading = true;
+    this.getStagesForApplication(selected.name)
+      .then((stages) => {
+        this.stages = stages;
+        this.viewState.loading = false;
+        this.viewState.error = false;
+      })
+      .catch(() => {
+        this.viewState.loading = false;
+        this.viewState.error = true;
+      });
+  }
+
+  cancel (): void {
+    this.$uibModalInstance.dismiss();
+  }
+
+  copyStage (): void {
+    this.$uibModalInstance.close(this.selectedStage.stage);
+  }
+}
+
+export const COPY_STAGE_MODAL_CONTROLLER = 'spinnaker.core.copyStage.modal.controller';
+
+module(COPY_STAGE_MODAL_CONTROLLER, [
+    API_SERVICE,
+    COPY_STAGE_CARD_COMPONENT,
+    require('core/application/service/applications.read.service.js'),
+  ])
+  .controller('CopyStageModalCtrl', CopyStageModalCtrl);
```

Comment:
It's a little pedantic, but can you add access modifiers (`public`, `private`) to fields and methods?